File: AdventOfCode.FSharp/2022/Day4.fs
namespace AdventOfCode.FSharp._2022

open AdventOfCode.Core
open AdventOfCode.FSharp
open Xunit

module Day4 =
    type Range = {
        From : int;
        To : int;
    }

    (*
     *  "2-4,6-8" -> [ "2-4"; "6-8" ] -> [ [ "2"; "4" ]; [ "6"; "8" ] ] -> [ [ 2; 4 ]; [ 6; 8 ] ] -> [ 2; 4; 6; 8 ]
     *
     *  [ 2; 4; 6; 8 ] -> { From = 2; To = 4 }, { From = 6; To = 8 }
     *)
    let Parse = 
        Shared.Split
        >> Seq.map (Shared.SplitBy "," >> Seq.map (Shared.SplitBy "-" >> Array.map int) >> Array.concat)
        >> Seq.map (fun a -> { From = a[0]; To = a[1] }, { From = a[2]; To = a[3]})

    (*
     *  r1  a...........b
     *  r2      c...d     
     *)
    let FullyContains (r1 : Range) (r2 : Range) = 
        r1.From <= r2.From && r2.To <= r1.To

    (*
     *  r1  a...b
     *  r2    c...d     
     *)
    let Overlaps (r1 : Range) (r2 : Range) = 
        r1.From <= r2.From && r2.From <= r1.To

    let Solve = 
        Parse
        >> Seq.filter (fun (r1, r2) -> FullyContains r1 r2 || FullyContains r2 r1)
        >> Seq.length

    let SolvePartTwo = 
        Parse
        >> Seq.filter (fun (r1, r2) -> Overlaps r1 r2 || Overlaps r2 r1)
        >> Seq.length

    let exampleInput =
        @"2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8"

    [<Fact>]
    let ``Solves Part One Example`` () =
        let actual = Solve exampleInput
        Assert.Equal(2, actual)

    [<Fact>]
    let ``Solves Part Two Example`` () =
        let actual = SolvePartTwo exampleInput
        Assert.Equal(4, actual)
        

type Day4(input: string) =
    interface ISolver with
        member this.SolvePartOne() = Day4.Solve input |> string
        member this.SolvePartTwo() = Day4.SolvePartTwo input |> string

File: AdventOfCode.CSharp/2023/Day4.cs
using Xunit;
using Xunit.Abstractions;

namespace AdventOfCode.CSharp._2023;

public class Day4 : Solver {
    public Day4(string? input = null, ITestOutputHelper? outputHelper = null) : base(input, outputHelper) { }

    public override long SolvePartOne() {
        var accumulator = 0;

        foreach (var line in Shared.Split(Input)) {
            var outerparts = line.Split(':'); 
            var parts = outerparts[1].Split('|');
            var winningNumbers = parts[0].Split(' ', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToArray();
            var myNumbers = parts[1].Split(' ', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToArray();

            var matches = myNumbers.Where(n => winningNumbers.Contains(n));
            var numberOfMatches = matches.Count();
            
            if (numberOfMatches > 0)
                accumulator += (int)Math.Pow(2, numberOfMatches - 1);
        }

        return accumulator;
    }

    public override long SolvePartTwo() {
        var lines = Shared.Split(Input);
        var accumulator = Enumerable.Repeat(1, lines.Length).ToArray();

        for (var i = 0; i < lines.Length; ++ i) {
            var line = lines[i];
            var outerparts = line.Split(':'); 
            var parts = outerparts[1].Split('|');
            var winningNumbers = parts[0].Split(' ', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToArray();
            var myNumbers = parts[1].Split(' ', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToArray();

            var matches = myNumbers.Where(n => winningNumbers.Contains(n));
            var numberOfMatches = matches.Count();

            if (numberOfMatches > 0) {
                for (var ii = i + 1; ii < i + 1 + numberOfMatches; ++ii) {
                    accumulator[ii] += accumulator[i];
                }
            }
        }

        return accumulator.Sum();
    }
    private const string? ExampleInput = @"
Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
";

    [Fact]
    public void SolvesPartOneExample() {
        var actual = new Day4(ExampleInput, Output).SolvePartOne();
        Assert.Equal(13, actual);
    }

    [Fact]
    public void SolvesPartTwoExample() {
        var actual = new Day4(ExampleInput, Output).SolvePartTwo();
        Assert.Equal(30, actual);
    }
}